File: Sources/PublicDocumentApp/OfficialStyleLint.swift
import Foundation

struct OfficialStyleLintFinding: Codable, Equatable {
    let ruleID: String
    let citation: String
    let message: String
    let dryRun: Bool
}

struct OfficialStyleLintReport: Codable, Equatable {
    let dryRun: Bool
    let findings: [OfficialStyleLintFinding]
}

enum OfficialStyleLint {
    private struct FilePayload: Decodable {
        struct Rule: Decodable {
            let id: String
            let kind: String
            let citation: String
            let message: String
        }
        let rules: [Rule]
    }

    static func dryRun(project: DocumentProject, catalog: TemplateCatalogPayload?) -> OfficialStyleLintReport {
        let rules = (try? loadRules()) ?? []
        let applied = catalog.flatMap { payload in
            payload.entries.first { $0.templateID == project.templateBinding.templateID }?.officialRules
        } ?? []
        let fields = Set(applied.map(\.field))
        let titleRule = applied.filter { $0.field == .title }.max { $0.precedence < $1.precedence }
        var findings: [OfficialStyleLintFinding] = []
        for rule in rules {
            let hit: Bool
            switch rule.kind {
            case "endMark":
                hit = fields.contains(.endMark) && !OfficialRuleSkeleton.hasEndMark(project.elements)
            case "attachment":
                hit = fields.contains(.attachment)
                    && OfficialRuleSkeleton.hasAttachments(project)
                    && !OfficialRuleSkeleton.hasAttachmentLabel(project.elements)
            case "senderName":
                hit = fields.contains(.senderName)
                    && !OfficialRuleSkeleton.isInternalApproval(project.elements)
                    && !OfficialRuleSkeleton.hasSenderName(project.elements)
            case "titleLock":
                hit = fields.contains(.title) && titleRule.map { project.title != $0.requiredValue } ?? false
            default:
                hit = false
            }
            if hit {
                findings.append(OfficialStyleLintFinding(
                    ruleID: rule.id,
                    citation: rule.citation,
                    message: rule.message,
                    dryRun: true
                ))
            }
        }
        return OfficialStyleLintReport(dryRun: true, findings: findings)
    }

    private static func loadRules() throws -> [FilePayload.Rule] {
        let bundled = Bundle.main.resourceURL?.appendingPathComponent("Rules/official-style-lint-1.0.0.json")
        let source = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
            .appendingPathComponent("Resources/Rules/official-style-lint-1.0.0.json")
        for candidate in [bundled, source].compactMap({ $0 }) where FileManager.default.fileExists(atPath: candidate.path) {
            return try JSONDecoder().decode(FilePayload.self, from: Data(contentsOf: candidate)).rules
        }
        return []
    }
}

File: public_document_web/official_style_lint.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, TypedDict

from .official_rules_skeleton import (
    has_attachment_label,
    has_attachments,
    has_end_mark,
    has_sender_name,
    highest_precedence_rules,
    is_internal_approval,
)
from .paths import RESOURCES

LINT_PATH = RESOURCES / "Rules" / "official-style-lint-1.0.0.json"


class LintReport(TypedDict):
    dryRun: bool
    findings: list[dict[str, Any]]


def load_lint_rules(path: Path | None = None) -> list[dict[str, Any]]:
    payload = json.loads((path or LINT_PATH).read_text(encoding="utf-8"))
    return list(payload.get("rules") or [])


def dry_run_lint(
    project: dict[str, Any],
    catalog: dict[str, Any] | None = None,
    path: Path | None = None,
) -> LintReport:
    findings: list[dict[str, Any]] = []
    elements = list(project.get("elements") or [])
    title_rule = None
    fields = {"endMark", "attachment", "senderName", "title"}
    if catalog is not None:
        binding = project.get("templateBinding") or {}
        applied = highest_precedence_rules(catalog, binding.get("templateID"))
        fields = {rule.get("field") for rule in applied}
        for rule in applied:
            if rule.get("field") == "title":
                title_rule = rule
                break
    for rule in load_lint_rules(path):
        kind = rule.get("kind")
        hit = False
        if kind == "endMark" and "endMark" in fields:
            hit = not has_end_mark(elements)
        elif kind == "attachment" and "attachment" in fields:
            hit = has_attachments(project) and not has_attachment_label(elements)
        elif kind == "senderName" and "senderName" in fields:
            hit = not is_internal_approval(elements) and not has_sender_name(elements)
        elif kind == "titleLock" and "title" in fields:
            hit = bool(title_rule) and project.get("title") != title_rule.get("requiredValue")
        if hit:
            findings.append({
                "ruleID": rule.get("id"),
                "citation": rule.get("citation"),
                "message": rule.get("message"),
                "dryRun": True,
            })
    return {"dryRun": True, "findings": findings}


def apply_lint(
    project: dict[str, Any],
    catalog: dict[str, Any] | None = None,
) -> tuple[dict[str, Any], dict[str, Any]]:
    report = dry_run_lint(project, catalog)
    locked = [item for item in report["findings"] if item.get("ruleID") == "lint-title-lock"]
    if locked:
        return project, {
            "dryRun": False,
            "applied": False,
            "findings": locked,
            "message": "공식 제목 덮어쓰기를 되돌리지 않습니다.",
        }
    return project, {
        "dryRun": False,
        "applied": False,
        "findings": report["findings"],
        "message": "린트 적용은 명시적 교정 경로가 아닙니다. dry_run 결과만 남깁니다.",
    }
